### PR TITLE
Update modals to use MediaLayout to implement mobile designs

### PR DESCRIPTION
**Pull request by kevinbarabash**
This PR builds on top of #372 by using MediaLayout to implement mobile modal designs.

TODOs: 
- [x] add some tests that use `overrideSize` to create snapshot tests for the mobile versions of the modals.

### CODE REVIEW

**Comment by jeresig** [comment]
Deploy preview for *wonder-blocks* ready!

Built with commit b460befa461efd6483564954ad22ff48a60d7b74

https://deploy-preview-373--wonder-blocks.netlify.com

File: packages/wonder-blocks-modal/components/close-button.js [hunk 1]
```diff
@@ -2,50 +2,39 @@
...
-     * tertiary button title to color.
-     */
-    light: boolean,
-
-    /**
-     * Called when the close button is clicked.
-     */
-    onClick?: () => void,
+    light?: boolean,
+    onClose?: () => void,
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/close-button.js]: Where are the prop comments and why the change to `onClose` from `onClick`. The `CloseButton` doesn't get closed, so `onClose` doesn't really make sense

File: packages/wonder-blocks-modal/components/one-column-modal.js [hunk 2]
```diff
@@ -51,53 +73,28 @@ class ContentWrapper extends React.Component<WrappedProps> {
...
+                <MediaLayout>
+                    {({mediaSize}) =>
+                        mediaSize === "small" ? (
+                            <SmallOneColumnModal {...this.props} />
+                        ) : (
+                            <LargeOneColumnModal {...this.props} />
+                        )
+                    }
+                </MediaLayout>
+            </React.Fragment>
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/one-column-modal.js]: Is the React.Fragment needed?

File: packages/wonder-blocks-modal/components/two-column-modal.js [hunk 3]
```diff
@@ -79,73 +126,32 @@ class ContentWrapper extends React.Component<WrappedProps> {
...
+                <MediaLayout>
+                    {({mediaSize}) =>
+                        mediaSize === "small" ? (
+                            <SmallTwoColumnModal {...this.props} />
+                        ) : (
+                            <LargeTwoColumnModal {...this.props} />
+                        )
+                    }
+                </MediaLayout>
+            </React.Fragment>
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/two-column-modal.js]: Is the React.Fragment necessary?

File: packages/wonder-blocks-modal/components/close-button.js (hunk 1 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-modal/components/close-button.js]: Good call.

File: packages/wonder-blocks-modal/components/one-column-modal.js (hunk 2 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-modal/components/one-column-modal.js]: I'll drop it.

**Comment by kevinbarabash** [comment]
I noticed I still have a TODO to add some fixture tests for the mobile versions of the modals so I'm going to add those now.

File: packages/wonder-blocks-modal/components/close-button.js [hunk 4]
```diff
@@ -2,34 +2,20 @@
...
-     * tertiary button title to color.
-     */
-    light: boolean,
-
-    /**
-     * Called when the close button is clicked.
-     */
+    light?: boolean,
     onClick?: () => void,
+    style?: StyleType,
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/close-button.js]: Shouldn't we still have comments to describe these props?

File: packages/wonder-blocks-modal/components/close-button.js [hunk 5]
```diff
@@ -38,14 +24,17 @@ export default class CloseButton extends React.Component<Props> {
...
                     }
+
                     return (
                         <IconButton
                             icon={icons.dismiss}
                             // TODO(mdr): Translate this string for i18n.
+                            // TODO(kevinb): provide a way to set this label
                             aria-label="Close modal"
-                            onClick={closeModal || onClick}
+                            onClick={onClick || closeModal}
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/close-button.js]: Why are these reversed?

File: packages/wonder-blocks-modal/components/modal-content.js [hunk 6]
```diff
@@ -26,54 +27,54 @@ export default class ModalContent extends React.Component<Props> {
...
-    small: {
-        padding: "32px 16px",
-    },
-});
+    small: StyleSheet.create({
+        content: {
+            padding: "32px 16px",
+        },
+    }),
+};
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/modal-content.js]: I like this pattern together with the `MediaLayout` component. Nice.

File: packages/wonder-blocks-modal/components/modal-dialog.js [hunk 7]
```diff
@@ -17,41 +17,48 @@ type Props = {|
 export default class ModalDialog extends React.Component<Props> {
     render() {
         const {style, children} = this.props;
+        const contextValue = {
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/modal-dialog.js]: Add the context type here so we can flow type this nicely?

File: packages/wonder-blocks-modal/components/modal-panel.js [hunk 8]
```diff
@@ -88,100 +85,113 @@ export default class ModalPanel extends React.Component<Props> {
...
-                {!footer ||
-                (typeof footer === "object" && footer.type === ModalFooter) ? (
-                    footer
-                ) : (
-                    <ModalFooter>{footer}</ModalFooter>
-                )}
-            </View>
+            <MediaLayout styleSheets={styleSheets}>
+                {({styles}) => {
+                    let mainContent =
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/modal-panel.js]: I find use of `let` to be a code smell that a function is needed. What do you think about using a function so that this can be `const` instead? Might make the code a little less nested and easier to follow.

File: packages/wonder-blocks-modal/components/one-column-modal.js [hunk 9]
```diff
@@ -27,22 +27,44 @@ type BaseProps = {|
...
+                <View style={styles.contentFooterWrapper}>
+                    <View style={styles.smallContentWrapper}>
+                        <ModalPanel
+                            showCloseButton
+                            onClose={onClose}
+                            content={content}
+                            scrollOverflow={false}
+                        />
+                    </View>
+                    {footer && (
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/one-column-modal.js]: Flow upgrades aren't going to like this. Change to `!!footer`.

File: packages/wonder-blocks-modal/components/one-column.modal.test.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+import {mount, unmountAll} from "../../../utils/testing/mount.js";
+
+describe("OneColumnModal", () => {
+    beforeEach(() => {
+        unmountAll();
+    });
+
+    it("should render SmallOneColumnModal on mobile", () => {
+        // Arrange
+        const wrapper = mount(
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/one-column.modal.test.js]: This is the action so it probably should go under `// Act`. Sometimes there isn't anything to arrange, and that's ok. Just have an empty `// Arrange` section.

File: packages/wonder-blocks-modal/components/one-column.modal.test.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        // Act
+        const modal = wrapper.find(LargeOneColumnModal);
+
+        // Assert
+        expect(modal);
+    });
+
+    // We already capture desktop snapshots from auto generated tests
+    describe("mobile", () => {
+        test("with footer", () => {
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/one-column.modal.test.js]: Although there's only one test case for this variation right now, it can be useful to use `describe` to layout general scenarios and then have cases within that.

Like `describe("with footer")` then have a test case like `it("should match snapshot")` or  `test("it matches snapshot")`.

The nice thing about that is that we then document the circumstances and the expectation, rather than just the circumstances as in the code as it stands.

What do you think?

File: packages/wonder-blocks-modal/components/standard-modal.js [hunk 12]
```diff
@@ -71,69 +88,65 @@ export default class StandardModal extends React.Component<Props> {
...
-                            //(mediaSize) =>
-                            //    mediaSize === "small" && styles.smallPreview
-                        ]}
-                        content={
-                            <ModalContent style={styles.previewContent}>
-                                {preview}
-                            </ModalContent>
-                        }
-                    />
+                        {preview && (
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/standard-modal.js]: `!!preview`

File: packages/wonder-blocks-modal/components/standard-modal.test.js [hunk 13]
```diff
@@ -2,14 +2,78 @@
...
+import {
+    MediaLayoutContext,
+    MEDIA_DEFAULT_SPEC,
+} from "@khanacademy/wonder-blocks-layout";
 import StandardModal from "./standard-modal.js";
+import {mount, unmountAll} from "../../../utils/testing/mount.js";
 
 describe("StandardModal", () => {
+    beforeEach(() => {
+        unmountAll();
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/standard-modal.test.js]: Thanks for add this!

File: packages/wonder-blocks-modal/components/standard-modal.test.js [hunk 14]
```diff
@@ -2,14 +2,78 @@
...
             <StandardModal title="Title" content="Content" footer="Footer" />,
         );
 
         expect(wrapper.find("ModalCloseButton").exists()).toBeFalsy();
     });
+
+    // We already capture desktop snapshots from auto generated tests
+    test("mobile", () => {
+        // Arrange
+        const wrapper = mount(
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/standard-modal.test.js]: This is really where the action is. What do you think about moving it to `// Act`?

File: packages/wonder-blocks-modal/components/two-column-modal.js [hunk 15]
```diff
@@ -34,29 +34,76 @@ type BaseProps = {|
...
+                            }
+                            scrollOverflow={false}
+                        />
+                        <ModalPanel
+                            style={styles.smallColumn}
+                            content={content}
+                            scrollOverflow={false}
+                        />
+                    </View>
+                    {footer && (
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/two-column-modal.js]: `!!footer`

Comment (somewhatabstract) [reply packages/wonder-blocks-modal/components/two-column-modal.js]: Or as with all the other cases, consider moving to a `maybeRenderFooter` method and then just have that return null when `footer` is falsy.

File: packages/wonder-blocks-modal/components/two-column-modal.test.js [hunk 16]
```diff
@@ -2,14 +2,104 @@
...
+            const wrapper = shallow(
+                <SmallTwoColumnModal
+                    content="Content"
+                    fullBleedSidebar={true}
+                    sidebar="Sidebar"
+                    footer="Footer"
+                />,
+            );
+
+            expect(wrapper).toMatchSnapshot();
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/two-column-modal.test.js]: These should still have the AAA layout, even if the first A is empty.

File: packages/wonder-blocks-modal/components/close-button.js (hunk 5 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-modal/components/close-button.js]: Not sure why I reversed them.  Only one is possible at a time since we throw an error if both are provided.

File: packages/wonder-blocks-modal/components/modal-content.js (hunk 6 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-modal/components/modal-content.js]: Glad to hear it.  I hope that other devs like it too. 🤞

File: packages/wonder-blocks-modal/components/one-column.modal.test.js (hunk 10 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-modal/components/one-column.modal.test.js]: In the case we're testing the rendering so having it under `// Act` makes sense.  I got in the habit from some of the previous tests putting the `mount` under arrange.

File: packages/wonder-blocks-modal/components/one-column.modal.test.js (hunk 11 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-modal/components/one-column.modal.test.js]: Sounds good.

File: packages/wonder-blocks-modal/components/close-button.js [hunk 17]
```diff
@@ -7,8 +7,18 @@ import type {StyleType} from "@khanacademy/wonder-blocks-core";
...
+     *
+     * Sets primary button background color to white, and secondary and
+     * tertiary button title to color.
+     */
     light?: boolean,
+
+    /** Optional click handler */
     onClick?: () => void,
+
+    /** Optional custom styles. */
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/close-button.js]: Thanks for adding these!

File: packages/wonder-blocks-modal/components/modal-dialog.js [hunk 18]
```diff
@@ -17,7 +18,7 @@ type Props = {|
 export default class ModalDialog extends React.Component<Props> {
     render() {
         const {style, children} = this.props;
-        const contextValue = {
+        const contextValue: MediaLayoutContextValue = {
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/modal-dialog.js]: Nice! (Not convinced we need the word `Value` on the end of that type, but I'm not holding a strong opinion on that).

File: packages/wonder-blocks-modal/components/modal-panel.js [hunk 19]
```diff
@@ -78,49 +79,53 @@ export default class ModalPanel extends React.Component<Props> {
...
+                ((content: any): React.Element<typeof ModalContent>)
+            ) : (
+                <ModalContent>{content}</ModalContent>
+            );
+
+        if (mainContent) {
+            return mainContent;
+        }
+
+        return React.cloneElement(mainContent, {
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/modal-panel.js]: If we get here, won't `mainContent` be `undefined` or otherwise falsy?

File: packages/wonder-blocks-modal/components/modal-panel.js [hunk 20]
```diff
@@ -78,49 +79,53 @@ export default class ModalPanel extends React.Component<Props> {
...
+        const mainContent =
+            content &&
+            typeof content === "object" &&
+            content.type === ModalContent ? (
+                ((content: any): React.Element<typeof ModalContent>)
+            ) : (
+                <ModalContent>{content}</ModalContent>
+            );
+
+        if (mainContent) {
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/modal-panel.js]: I think this should be `!mainContent`.

File: packages/wonder-blocks-modal/components/modal-panel.js [hunk 21]
```diff
@@ -78,49 +79,53 @@ export default class ModalPanel extends React.Component<Props> {
...
-                                !!titleBar && styles.hasTitleBar,
-                                !!footer && styles.hasFooter,
-                                showCloseButton &&
-                                    !titleBar &&
-                                    styles.withCloseButton,
-                                mainContent.props.style,
-                            ],
-                        });
-                    }
+                    const mainContent = this.renderMainContent(styles);
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/modal-panel.js]: Nice.

File: packages/wonder-blocks-modal/components/one-column.modal.test.js [hunk 22]
```diff
@@ -61,18 +61,24 @@ describe("OneColumnModal", () => {
...
-            const wrapper = shallow(<SmallOneColumnModal content="Content" />);
+        describe("without footer", () => {
+            it("should match snapshot", () => {
+                const wrapper = shallow(
+                    <SmallOneColumnModal content="Content" />,
+                );
 
-            expect(wrapper).toMatchSnapshot();
+                expect(wrapper).toMatchSnapshot();
+            });
```
Comment (somewhatabstract) [inline packages/wonder-blocks-modal/components/one-column.modal.test.js]: Think these should follow the AAA comment format.

File: packages/wonder-blocks-modal/components/modal-dialog.js (hunk 18 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-modal/components/modal-dialog.js]: I'm not super happy about the name either but `MediaLayoutContext` is already taken. 😞
